Restrict performance entries from no-cors cross-origin media requests

sysrqb opened the pull request:
#### 0473037b55025aebdf5a4e9297518b832ab3aaac
<pre>
Restrict performance entries from no-cors cross-origin media requests
<a href="https://bugs.webkit.org/show_bug.cgi?id=276208">https://bugs.webkit.org/show_bug.cgi?id=276208</a>
<a href="https://rdar.apple.com/124946839">rdar://124946839</a>

Reviewed by Ryosuke Niwa.

This patch now sets a media request as coming from an opaque source if it is
  1. cross-origin
  2. no-cors
  3. a nested resource (e.g., from an index file)

* LayoutTests/http/tests/media/resources/hls/.htaccess: Added.
* LayoutTests/http/tests/performance/performance-resource-timing-cross-origin-media-expected.txt: Added.
* LayoutTests/http/tests/performance/performance-resource-timing-cross-origin-media-with-cors-expected.txt: Added.
* LayoutTests/http/tests/performance/performance-resource-timing-cross-origin-media-with-cors.html: Added.
* LayoutTests/http/tests/performance/performance-resource-timing-cross-origin-media.html: Added.
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/wincairo/TestExpectations:
* Source/WebCore/loader/MediaResourceLoader.cpp:
(WebCore::MediaResourceLoader::requestResource):
* Source/WebCore/loader/MediaResourceLoader.h:

Canonical link: <a href="https://commits.webkit.org/280985@main">https://commits.webkit.org/280985@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/965daadca97d10b16b2b1155cbf865336aa3225a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/58230 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/37558 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/10713 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/61855 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/8675 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/60359 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/45194 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/8869 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/47187 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/6200 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/60261 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/35209 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/50347 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/28021 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/31972 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/7644 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/7679 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/53927 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/7913 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/63560 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/2144 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/7975 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/54505 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/2152 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/50359 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/54564 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/1837 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8686 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/33387 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/34473 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/35557 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/34218 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/loader/ResourceTimingInformation.cpp:
```diff
@@ -36,19 +36,20 @@
     return resource.resourceRequest().url().protocolIsInHTTPFamily()
         && !resource.loadFailedOrCanceled()
-        && resource.options().loadedFromOpaqueSource == LoadedFromOpaqueSource::No;
+        && (resource.options().loadedFromOpaqueSource == LoadedFromOpaqueSource::No || passesTimingAllowOriginCheck(resource.response(), documentOrigin));
```
Please explain why we're making this change in the commit message.

rniwa commented on Source/WebKit/WebProcess/Network/WebLoaderStrategy.cpp:
```diff
@@ -415,7 +415,8 @@ void WebLoaderStrategy::scheduleLoadFromNetworkProcess(ResourceLoader& resourceL
     addParametersShared(frame, loadParameters, isMainFrameNavigation);
 
-    loadParameters.serviceWorkersMode = resourceLoader.options().loadedFromOpaqueSource == LoadedFromOpaqueSource::No ? resourceLoader.options().serviceWorkersMode : ServiceWorkersMode::None;
+    auto shouldSkipServiceWorker = resourceLoader.options().loadedFromOpaqueSource == LoadedFromOpaqueSource::Yes && resourceLoader.cachedResource() && resourceLoader.cachedResource()->type() != CachedResource::Type::MediaResource;
```
Why are we explicitly allowing media resources to be processed by the service worker?
The rationale should be explained in the commit message.

sysrqb replied:
It's not clear to me why we currently prohibit loading CSS from service workers. Is this defined in a spec? For media, WPT implies that this should be allowed, so that's what I'm following. Chrome currently passes while Firefox is timing out (and WebKit times-out if we prohibit service workers): https://wpt.fyi/results/service-workers/service-worker/fetch-request-resources.https.html?label=master&label=experimental&aligned

cc @youennf, too, for thoughts about how we should behave here.

sysrqb committed:
Restrict performance entries from no-cors cross-origin media requests
https://bugs.webkit.org/show_bug.cgi?id=276208
rdar://124946839

Reviewed by Ryosuke Niwa.

This patch now sets a media request as coming from an opaque source if it is
  1. cross-origin
  2. no-cors
  3. a nested resource (e.g., from an index file)

* LayoutTests/http/tests/media/resources/hls/.htaccess: Added.
* LayoutTests/http/tests/performance/performance-resource-timing-cross-origin-media-expected.txt: Added.
* LayoutTests/http/tests/performance/performance-resource-timing-cross-origin-media-with-cors-expected.txt: Added.
* LayoutTests/http/tests/performance/performance-resource-timing-cross-origin-media-with-cors.html: Added.
* LayoutTests/http/tests/performance/performance-resource-timing-cross-origin-media.html: Added.
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/wincairo/TestExpectations:
* Source/WebCore/loader/MediaResourceLoader.cpp:
(WebCore::MediaResourceLoader::requestResource):
* Source/WebCore/loader/MediaResourceLoader.h:

Canonical link: https://commits.webkit.org/280985@main
